### PR TITLE
Migrate Core Binding XML Tests from Groovy to Java

### PR DESCRIPTION
**Pull request by wborn**
Migrates the Core Binding XML Tests from Groovy to Java.

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        ConfigurationAdmin configAdmin = getService(ConfigurationAdmin.class);
+        Configuration config = configAdmin.getConfiguration("org.eclipse.smarthome.core.i18nprovider", null);
+        Dictionary<String, String> localeCfg = new Hashtable<>();
+        localeCfg.put("language", "de");
+        localeCfg.put("country", "DE");
+        config.update(localeCfg);
+
+        // before running the test with a default locale make sure the locale has been set
+        LocaleProvider localeProvider = getService(LocaleProvider.class);
+        waitForAssert(new Closure<Object>(null) {
```
Comment (wborn) [inline bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java]: Looks like you currently can only use Groovy Closures with waitForAssert in the OSGiTest. There are no methods with Suppliers like in the JavaTest.

File: bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+public class BindingInfoI18nTest extends JavaOSGiTest {
+
+    private static final String TEST_BUNDLE_NAME = "yahooweather.bundle";
+
+    private BindingInfoRegistry bindingInfoRegistry;
+
+    @Before
+    public void setUp() {
+        bindingInfoRegistry = getService(BindingInfoRegistry.class);
+        assertThat(bindingInfoRegistry, is(notNullValue()));
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java]: JUnit Assert: `assertNotNull(bindingInfoRegistry);`

File: bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        SyntheticBundleInstaller.uninstall(bundleContext, TEST_BUNDLE_NAME);
+    }
+
+    @Test
+    public void assertBindingInfosWereLocalizedInGerman() throws Exception {
+        int initialNumberOfBindingInfos = bindingInfoRegistry.getBindingInfos().size();
+
+        // install test bundle
+        Bundle bundle = SyntheticBundleInstaller.install(bundleContext, TEST_BUNDLE_NAME);
+        assertThat(bundle, is(notNullValue()));
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java]: JUnit Assert: `assertNotNull(bundle);`

File: bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        Set<BindingInfo> bindingInfos = bindingInfoRegistry.getBindingInfos(Locale.GERMAN);
+        assertThat(bindingInfos.size(), is(initialNumberOfBindingInfos + 1));
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test/java/org/eclipse/smarthome/core/binding/xml/test/BindingInfoI18nTest.java]: JUnit Assert: `assertEquals(initialNumberOfBindingInfos + 1, bindingInfos.size());`

**Comment by maggu2810** [comment]
>
> Looks like you currently can only use Groovy Closures with waitForAssert
> in the OSGiTest. There are no methods with Suppliers like in the JavaTest.
>

Hm, there exist a few `waitForAssert` that are using `Runnable` and
`Supplier` as the first argument. Best candidates for closures.

**Comment by wborn** [comment]
The issue with the Groovy closures was that the test had to extend JavaOSGiTest instead of OSGiTest. :-)

**Comment by wborn** [comment]
> I need to have a look at, but IIRC there has been some comments of @sjka and mine about the not needed Hamcrest matchers some time ago.
>
> -- edit--
> 
> #5166 (review)

If Hamcrest should no longer be used (i.e. like Groovy). It should be clearly stated in the documentation.

It currently promotes the use of Hamcrest: :wink: 

https://www.eclipse.org/smarthome/documentation/development/testing.html

> Using the the [https://code.google.com/p/hamcrest/ hamcrest] matcher library is a good way to write  expressive assertions. In contrast to the original assertion statements from JUnit the hamcrest matcher library allows to define the assertion in a more natural order:
> 
> PercentType pt = new PercentType("0.0001");
> assertThat(pt.toString(), is(equalTo("0.0001")));

**Comment by maggu2810** [comment]
You are right, the `testing.md` file and so the generated documentation advise to use Hamcrest.
IIRC that documentation (at least using `git blame`) is mostly written by @kaikreuzer and @htreu so if they prefer to use that dependency because of more "expressive assertions" I don't care about.

My personal feeling is that a developer should be able to read `assertEquals("0.0001", pt.toString());` and `assertThat(pt.toString(), is(equalTo("0.0001")));` the same way. And before I need to learn the Hamcrest API just to use other statements I prefer to stay at JUnit itself. :wink: 

So only my personal feeling, you can keep the Hamcrest staff.

**Comment by wborn** [comment]
Perhaps the documentation may need an update because of progressive insights and being able to use Java 8 nowadays. It helps to get an updated view on if/how Hamcrest should be used since it seems to be both promoted and discouraged at the same time. I've most of the time also just used JUnit without any Hamcrest at all. :slightly_smiling_face: 

With these migration PRs my main focus is to get rid of Groovy tests. If it is just a few statements that need to be adjusted to also remove the Hamcrest dependency it's not much additional work.

**Comment by kaikreuzer** [comment]
> And before I need to learn the Hamcrest API just to use other statements I prefer to stay at JUnit itself. 

Same for me - I am not a fan of Hamcrest and fine to have the tests without it. So yes, it might make sense to remove it from the docs and also not include it the test archetype as a dependency anymore. @htreu What's your take?

**Comment by htreu** [comment]
Imho Hamcrest adds some useful assertion tools, especially when dealing with collections and arrays like `assertThat(myArray, arrayContainig("item1", "item2"))`.
Since hamcrest is a test only dependency my suggestion is to keep it in the project and use it whenever useful and/or when the developer writing the test feels comfortable using it. In this case we should not enforce a strict rule but leave some room for personal preferences.
wdyt?

**Comment by maggu2810** [comment]
If assertions are more readable -- I don't think of a native human readable code but the error message if the assertion fails -- or if the code is much simpler with Hamcrest then with plain JUnit (I trust @htreu that it is the case for e.g. collections and arrays) then Hamcrest can be used (IMHO).

But I don't think the usage should be such prominent and we could remove it from the docs and the archetype.

Could we agree on that:
* remove Hamcrest usage from doc
* remove Hamcrest dependencies from archetype
* allow to require developers in review to use `assertNotNull`, `assertNull`, `assertEquals`, `assertTrue`, ... for simple statements in PRs
* allow the developers to use `assertThat` for more complex matches

**Comment by htreu** [comment]
Just to be clear:

> allow to require developers in review to use ....

means its up to the contributor/reviewer constellation on whether the usage is okay? Or do you aim for a consistent way and `simple` & `complex` should be strictly defined?

> remove Hamcrest usage from doc, remove Hamcrest dependencies from archetype

why? Either we allow the usage of Hamcrest as a test framework or we don't. If we agree on using it, then why hide it? Imho it eases the entry into writing expressive tests.
In case we agree on not allowing it we should not except further contributions using it.

I suggest to not take away and discourage the use of a test dependency which provides more expressive freedom. Especially when we try to encourage contributors to provide tests Hamcrest makes it less painful to write them.
When it comes to the discussion on discontinuing groovy there are some concrete technical arguments against it (like: access to private members/methods leads to bad API design, bad refactoring support, etc.). I don't see such technical arguments against Hamcrest.

**Comment by maggu2810** [comment]
> Either we allow the usage of Hamcrest as a test framework or we don't.

For me it is not only black or white. IMHO it is the same as generics, streams, optionals, ... All that are very great language features, but using only that constructs for every single task is not necessary and does not make sense.

You are right, we should not remove it from the docs, we should still state it is allowed.
But can't we add a better example?

> Using the [https://code.google.com/p/hamcrest/ hamcrest] matcher library is a good way to write expressive assertions.

with an example that demonstrates that `assertThat(pt.toString(), is(equalTo("0.0001")));` could be used instead of `assertEquals("0.0001", pt.toString());` sounds to me that `assertThat` should be used for every simple assertion.

It is (for me) like a statement that we support streams and explicitly state we could use `IntStream.range(...)` instead of `for (int i = ...)`.

>> allow to require developers in review to use ....

> means its up to the contributor/reviewer constellation on whether the usage is okay? Or do you aim for a consistent way and `simple` & `complex` should be strictly defined?

My intention about the examples I added after my quoted message has been that for simple assertion it would perhaps make more sense to keep the old assertion that are still part of the JUnit API.

What`s wrong writing `assertTrue(conditionResult)`?


As stated above I have not used Hamcrest but starting with the latest PRs I have had a short look at that usage,

If we look at the error message here:

```
assertTrue(result instanceof String);
// error message:
java.lang.AssertionError
    at org.junit.Assert.fail(Assert.java:86)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertTrue(Assert.java:52)
// ...
```

```
assertThat(result, instanceOf(String.class));
// error message:
java.lang.AssertionError:
Expected: an instance of java.lang.String
     but: null
    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
// ...
```

Hamcrest makes really sense. :wink: 

So, also okay for me to keep the situation as it is...

@htreu Can you give me your oppinion about
* `assertTrue(variable)` (same for false)
* `assertNotNull(variable)` (same for null)

does the usage of Hamcrest brings any win in that situation?

**Comment by htreu** [comment]
Thanks @maggu2810 for the feedback.

For all assertions done by Hamcrest's `assertThat` we would profit from the advanced output you already mentioned above.
From a coding pov I see no big win compared to the build in JUnit assertion (assertTrue, assertNotNull, etc). Only for situations where `assertTrue` is used to assert none boolean functional aspects like `assertTrue(myList.size() == 3)` I would even in a review request a change. With Hamcrest this would result in:

```
assertThat(Arrays.asList("one", "two", "three"), hasSize(2));
// error message
java.lang.AssertionError: 
Expected: a collection with size <2>
     but: collection size was <3>
...
```

This helps understanding failing tests a lot. So I would also like to see a co-existence of JUnit assertions and Hamcrest with some freedom of choice for the developer.

I'm glad if we can agree on keep on using Hamcrest. Especially for none-trivial assertions it adds quite some benefit.

> But can't we add a better example?

We should! I will inspect our archetype and see if more/better examples can be added.

**Pull request merged**